src/liballoc/heap.rs

```diff
@@ -44,6 +45,84 @@ fn check_size_and_alignment(size: usize, align: usize) {
                   align);
 }
 
+#[derive(Copy, Clone, Default, Debug)]
+pub struct HeapAllocator;
+
+unsafe impl Allocator for HeapAllocator {
+    unsafe fn alloc(&mut self, layout: Layout) -> Result<*mut u8, AllocErr> {
```

Comment:
I am trying to ensure that the method is a minimal wrapper around the heap methods, reducing the chance that introducing the `Allocator` API causes regression in code speed or size...

also, re-reading the discussion on [the RFC](https://github.com/rust-lang/rfcs/pull/1398#issuecomment-204485595), there was some pushback saying that the `_unchecked` method variants should be removed.

I.e. I may have made a mistake in including them in this PR... but then again, I'm not sure if there was consensus about removing those methods.

In any case, if the `_unchecked` methods go away, then that seems like all the more reason to ensure that the base methods like `fn alloc` remain as lean as possible, even if that means making them `unsafe` and forcing the caller to ensure things like non-zero sized input layout...